crates/typst/src/eval/version.rs

```diff
@@ -0,0 +1,223 @@
+            None => bail!("unknown version component"),
+        }
+    }
+
+    /// Get an iterator of the values of all named version components for a version.
+    pub fn named_components(&self) -> impl Iterator<Item = (&'static str, i64)> + '_ {
+        Self::COMPONENT_NAMES
+            .into_iter()
+            .enumerate()
+            .map(|(i, name)| (name, self.get(i).unwrap_or_default().get()))
+    }
+
+    /// Convert a version into an array
+    pub fn into_array(self) -> Array {
+        self.0.into_iter().map(|i| Value::Int(i.get())).collect()
+    }
+
+    pub fn push(&mut self, component: VersionComponent) {
+        self.0.push(component);
+    }
+}
+
+impl FromIterator<VersionComponent> for Version {
+    fn from_iter<T: IntoIterator<Item = VersionComponent>>(iter: T) -> Self {
+        Self(EcoVec::from_iter(iter))
+    }
+}
+
+impl Ord for Version {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let max_len = self.0.len().max(other.0.len());
+        let tail = repeat(&VersionComponent::ZERO);
+
+        let self_iter = self.0.iter().chain(tail.clone());
+        let other_iter = other.0.iter().chain(tail);
+
+        for (l, r) in self_iter.zip(other_iter).take(max_len) {
+            match l.cmp(r) {
+                Ordering::Equal => (),
+                ord => return ord,
+            }
+        }
+        Ordering::Equal
+    }
+}
+impl PartialOrd for Version {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Eq for Version {}
+impl PartialEq for Version {
+    fn eq(&self, other: &Self) -> bool {
+        matches!(self.cmp(other), Ordering::Equal)
+    }
+}
+
+impl Hash for Version {
+    // Don't hash any explicitly suffixed zeros so that equal but non-identical versions (e.g. 1.2 and 1.2.0) hash as the same version
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let len = self
+            .0
+            .iter()
+            .rposition(|x| x.get() != 0)
+            .unwrap_or_else(|| self.0.len());
+
+        self.0[..len].hash(state);
+    }
+}
+
+impl Display for Version {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        let mut first = true;
+        for &i in &self.0 {
+            if first {
+                first = false;
+            } else {
+                write!(f, ".")?;
+            }
+            write!(f, "{i}")?;
+        }
+        Ok(())
+    }
+}
+
+impl Debug for Version {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "version({})", self)
+    }
+}
+
+impl Serialize for Version {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
+#[derive(Default, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+#[repr(transparent)]
+pub struct VersionComponent(u64);
+
+#[allow(dead_code)] // reason: kept for completeness, might be used in the future
```

Comment:
generally, I prefer not to have dead code. if we need something in the future, we can add it.